tf2 Tutorial: Timestamp and time travel tf2 tutorials (Python) (#1770)
ros2/ros2_documentation: source/Tutorials/Tf2/Learning-About-Tf2-And-Time-Py.rst

.. _LearningAboutTf2AndTimePy:

Learning about tf2 and time (Python)
====================================

**Goal:** Learn to use the ``timeout`` in ``lookup_transform`` function to wait for a transform to be available on the tf2 tree.

**Tutorial level:** Intermediate

**Time:** 10 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

> In previous tutorials, we recreated the turtle demo and learned to add a new frame to the transformation tree.

Reviewer: some links to previous tutorials

Author: I will add all links when we merge previous PR and I will resolve conflicts

.. _LearningAboutTf2AndTimePy:

Learning about tf2 and time (Python)
====================================

**Goal:** Learn to use the ``timeout`` in ``lookup_transform`` function to wait for a transform to be available on the tf2 tree.

**Tutorial level:** Intermediate

**Time:** 10 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

In previous tutorials, we recreated the turtle demo by writing a :ref:`tf2 broadcaster <WritingATf2BroadcasterPy>` and a :ref:`tf2 listener <WritingATf2ListenerPy>`.
We also learned how to :ref:`add a new frame to the transformation tree <AddingAFramePy>`.

> This tutorial will teach you how use the timeout in ``lookup_transform`` function to wait for a transform to be available on the tf2 tree.

Reviewer 2: It would be nice to list a few specific examples of when timeouts might be useful.

Author: Added in https://github.com/ros2/ros2_documentation/pull/1770/commits/a5227441dbabac6a69a1161c511388d11f02491c

Tasks
-----

1 Update the listener node
^^^^^^^^^^^^^^^^^^^^^^^^^^

Edit ``turtle_tf2_listener.py`` and remove the ``timeout=Duration(seconds=1.0)`` parameter that is passed to the ``lookup_transform()`` call on line 76.
It should look like shown below:

.. code-block:: python

   trans = self._tf_buffer.lookup_transform(
      to_frame_rel,
      from_frame_rel,
      now)

Moreover, import additional exceptions that we will handle in the beggining of the file:

.. code-block:: python

   from tf2_ros import LookupException, ConnectivityException, ExtrapolationException

Edit the exception handling on line 81 by adding newly imported exceptions and ``raise`` statement to see the exception:

.. code-block:: python

   except (LookupException, ConnectivityException, ExtrapolationException):
      self.get_logger().info('transform not ready')
      raise
      return

Now run the launch file:

.. code-block:: console

    ros2 launch learning_tf2_py turtle_tf2_demo.launch.py

> 

Reviewer 2: Since you know that this is not going to work, it would be nice to mention this. Perhaps move the first part of the next section up here.

When I am following tutorials, sometimes I'll try to get a section working before proceeding. It is frustrating when the section wasn't intended to work and didn't mention it.

Author: Solved in https://github.com/ros2/ros2_documentation/pull/1770/commits/ea33f90d351769517e9bcf721582547649e82866

2 Fix the listener node
^^^^^^^^^^^^^^^^^^^^^^^

> So, all of the sudden ``lookup_transform()`` is failing, telling you that the frame does not exist or that the data is in the future.

Reviewer 2: I prefer fewer words and shorter sentences.
```suggestion
Now ``lookup_transform()`` is failing. It tells you that the frame does not exist or that the data is in the future.
```

Author: Solved in https://github.com/ros2/ros2_documentation/pull/1770/commits/ea33f90d351769517e9bcf721582547649e82866

To fix this, edit your code on line 76 as shown below (return the ``timeout`` parameter):

.. code-block:: python

   trans = self._tf_buffer.lookup_transform(
      to_frame_rel,
      from_frame_rel,
      now,
      timeout=Duration(seconds=1.0))

The ``lookup_transform`` can take four arguments, where the last one is an optional timeout.
It will block for up to that duration waiting for it to timeout.

.. note::

   Once this change is made, remove the ``raise`` line from the ``except()`` block that we added above or the code will continue to fail.

You can now run the launch file.

.. code-block:: console

   ros2 launch learning_tf2_py turtle_tf2_demo.launch.py

You should notice that ``lookup_transform()`` will actually block until the transform between the two turtles becomes available (this will usually take a few milli-seconds).

> Once the timeout has been reached (1 second in this case), an exception will be raised only if the transform is still not available.

Reviewer 2: I think it is best to [write out nonnegative integers less than 100 or so](https://www.grammarly.com/blog/when-to-spell-out-numbers/).
```suggestion
Once the timeout has been reached (one second in this case), an exception will be raised only if the transform is still not available.
```

Author: Solved in https://github.com/ros2/ros2_documentation/pull/1770/commits/ea33f90d351769517e9bcf721582547649e82866
